libcudacxx/include/cuda/__memory_resource/resource_ref.h

```diff
@@ -68,79 +127,180 @@ struct _Async_alloc_vtable : public _Alloc_vtable
   _AsyncDeallocFn __async_dealloc_fn;
 
   constexpr _Async_alloc_vtable(
+    bool __is_small_,
     _Alloc_vtable::_AllocFn __alloc_fn_,
     _Alloc_vtable::_DeallocFn __dealloc_fn_,
     _Alloc_vtable::_EqualFn __equal_fn_,
+    _Alloc_vtable::_DestroyFn __destroy_fn_,
+    _Alloc_vtable::_MoveFn __move_fn_,
+    _Alloc_vtable::_CopyFn __copy_fn_,
     _AsyncAllocFn __async_alloc_fn_,
     _AsyncDeallocFn __async_dealloc_fn_) noexcept
-      : _Alloc_vtable(__alloc_fn_, __dealloc_fn_, __equal_fn_)
+      : _Alloc_vtable(__is_small_, __alloc_fn_, __dealloc_fn_, __equal_fn_, __destroy_fn_, __move_fn_, __copy_fn_)
       , __async_alloc_fn(__async_alloc_fn_)
       , __async_dealloc_fn(__async_dealloc_fn_)
   {}
 };
 
-// clang-format off
 struct _Resource_vtable_builder
 {
-    template <class _Resource, class _Property>
-    static __property_value_t<_Property> _Get_property(void* __res) noexcept {
-        return get_property(*static_cast<const _Resource *>(__res), _Property{});
-    }
+  template <_WrapperType _Wrapper_type>
+  using __wrapper_type = _CUDA_VSTD::integral_constant<_WrapperType, _Wrapper_type>;
 
-    template <class _Resource>
-    static void* _Alloc(void* __object, size_t __bytes, size_t __alignment) {
-        return static_cast<_Resource *>(__object)->allocate(__bytes, __alignment);
-    }
+  template <class _Resource, class _Property>
+  static __property_value_t<_Property> _Get_property(void* __res) noexcept
+  {
+    return get_property(*static_cast<const _Resource*>(__res), _Property{});
+  }
 
-    template <class _Resource>
-    static void _Dealloc(void* __object, void* __ptr, size_t __bytes, size_t __alignment) {
-        return static_cast<_Resource *>(__object)->deallocate(__ptr, __bytes, __alignment);
-    }
+  template <class _Resource>
+  static void* _Alloc(void* __object, size_t __bytes, size_t __alignment)
+  {
+    return static_cast<_Resource*>(__object)->allocate(__bytes, __alignment);
+  }
 
-    template <class _Resource>
-    static void* _Alloc_async(void* __object, size_t __bytes, size_t __alignment, ::cuda::stream_ref __stream) {
-        return static_cast<_Resource *>(__object)->allocate_async(__bytes, __alignment, __stream);
-    }
+  template <class _Resource>
+  static void _Dealloc(void* __object, void* __ptr, size_t __bytes, size_t __alignment) noexcept
+  {
+    // TODO: this breaks RMM because their memory resources do not declare their
+    // deallocate functions to be noexcept. Comment out the check for now until
+    // we can fix RMM.
```

Comment:
> I suppose it is because deallocation might often be called by destructors in an RAII pattern and those should not throw?

that's the thinking, yeah.

> we might observe cuda API errors in deallocation which we can then not throw

for that reason, when we call CUDA APIs in destructors and such, we typically use `_CUDA_ASSERT_CUDA_API` instead of the potentially-throwing `_CUDA_TRY_CUDA_API`.